### PR TITLE
New singularity and joint limit avoidance method

### PR DESCRIPTION
**Pull request by bbrito**

### CODE REVIEW

**Review by fmessmer — CHANGES_REQUESTED** [review]
this PR is not cleanly separated from other features...please rebase to latest `ipa320/indigo_dev` but maybe hold until we merged #148

**Comment by fmessmer** [comment]
also, you seem to use two different github user accounts...maybe you want to check your gitconfig files...

**Comment by fmessmer** [comment]
@ipa-bfb now that #150 is merged, please rebase this PR

**Comment by bbrito** [comment]
Rebased... Let me know your comments...

**Review by fmessmer — CHANGES_REQUESTED** [review]
without looking into it too closely, I already found few things you already should know:
 - indentation!
 - do not change default values!
 - do not commit strange files (test/.idea/* files)
 - do not commit all the test files...they are to specific to merge to `ipa320/indigo_dev`...keep them on a test_branch of  yours...

**Comment by bbrito** [comment]
Well about the default values I think we could make this solver the default at the moment since is the one with the best stable behaviour despite it does not avoid self-collisions... What do you say?

**Comment by bbrito** [comment]
@ipa-fxm could you give me some feedback here?

File: cob_twist_controller/scripts/test/test_careobot_st_jla_ca_sphere.py [hunk 1]
```diff
@@ -91,20 +91,21 @@ def init_pos():
...
+    if rospy.has_param(action_name+'chain_tip_link'):
         chain_tip_link = rospy.get_param('chain_tip_link')
     else:
         rospy.logwarn('Could not find parameter chain_tip_link.')
         exit(-1)
 
-    if rospy.has_param('frame_tracker/tracking_frame'):
-        tracking_frame = rospy.get_param('frame_tracker/tracking_frame')
+    if rospy.has_param(action_name+'frame_tracker/target_frame'):
+        tracking_frame = rospy.get_param(action_name+'frame_tracker/target_frame')
```
Comment (bbrito) [inline cob_twist_controller/scripts/test/test_careobot_st_jla_ca_sphere.py]: if you run this node like `rosrun cob_twist_controller test_careobot_st_jla_ca_sphere.py __ns:=arm_left`, you do not need these changes...

File: cob_twist_controller/src/cob_twist_controller.cpp [hunk 2]
```diff
@@ -298,6 +302,14 @@ void CobTwistController::checkSolverAndConstraints(cob_twist_controller::TwistCo
         warning = true;
     }
 
+    if (UNIFIED_JLA_SA == solver && CA_OFF != static_cast<ConstraintTypesCA>(config.constraint_ca))
+    {
+        ROS_ERROR("The Unified JLA and SA solution doesn\'t support collision avoidance. Currently UNIFIED_JLA_SA is only implemented for Identity and JLA ...");
```
Comment (bbrito) [inline cob_twist_controller/src/cob_twist_controller.cpp]: adjust comment

File: cob_twist_controller/src/constraint_solvers/constraint_solver_factory.cpp [hunk 3]
```diff
@@ -90,15 +91,20 @@ bool ConstraintSolverFactory::getSolverFactory(const TwistControllerParams& para
...
+
+        	    case JLA_OFF:
+        	        solver_factory.reset(new SolverFactory<WeightedLeastNormSolver>(params, limiter_params, task_stack_controller));
+        	        break;
             }
-            break;
+        	break;
+        case UNIFIED_JLA_SA:
+        	 solver_factory.reset(new SolverFactory<UnifiedJointLimitSingularitySolver>(params, limiter_params, task_stack_controller));
+        	 break;
```
Comment (bbrito) [inline cob_twist_controller/src/constraint_solvers/constraint_solver_factory.cpp]: fix indentation

**Review by fmessmer — APPROVED** [review]

**Pull request merged**